src/mqt/predictor/rl/actions/bqskit_actions.py

```
@@ -270,33 +477,83 @@ def final_layout_bqskit_to_qiskit(
-    else:
-        msg = f"Unhandled BQSKit pass type: {action.pass_type}"
-        raise ValueError(msg)
 
-    return bqskit_to_qiskit(bqskit_compiled_qc), layout
+    # ROUTING actions require existing layout
+    if action.pass_type == PassType.ROUTING:
+        assert layout is not None, "BQSKit routing requires an existing layout."
+        factory = cast("Callable[[Target], Callable[[Circuit], BQSKitMapping]]", action.transpile_pass)
+        compiled_qc, _initial, final = factory(device)(bqskit_qc)
+        compiled_qiskit_qc = bqskit_to_qiskit(compiled_qc)
+        output_qubits = layout._output_qubit_list  # ruff:ignore[private-member-access]
```

Ruff (0.16.1)

[warning] Invalid rule code in suppression: private-member-access

Enable `lint.preview` to use rule names

(RUF102)